### ISSUE TITLE
SQL Error on Staging Server

### ISSUE THREAD

**Issue by dylan-k** [WaltersArtMuseum/manuscripts-thewalters-org#82]
After running the site on latest stable Ubuntu, PHP, MySQL, we see some syntax errors.

```
        PHP Fatal error:  Uncaught PDOException: SQLSTATE[42000]: Syntax error or access violation: 1055 Expression #1 of SELECT list is not in GROUP BY clause and contains nonaggregated column 'manuscripts.folios.thumb_url' which is not functionally dependent on columns in GROUP BY clause; this is incompatible with sql_mode=only_full_group_by in /var/www/manuscriptsdev/home.php:172
Stack trace:
#0 /var/www/manuscriptsdev/home.php(172): PDOStatement->execute()
#1 /var/www/manuscriptsdev/index.php(183): include('...')
#2 {main}
  thrown in /var/www/manuscriptsdev/home.php on line 172

Fatal error: Uncaught PDOException: SQLSTATE[42000]: Syntax error or access violation: 1055 Expression #1 of SELECT list is not in GROUP BY clause and contains nonaggregated column 'manuscripts.folios.thumb_url' which is not functionally dependent on columns in GROUP BY clause; this is incompatible with sql_mode=only_full_group_by in /var/www/manuscriptsdev/home.php:172
Stack trace:
#0 /var/www/manuscriptsdev/home.php(172): PDOStatement->execute()
#1 /var/www/manuscriptsdev/index.php(183): include('...')
#2 {main}
        PHP Fatal error:  Uncaught PDOException: SQLSTATE[42000]: Syntax error or access violation: 1055 Expression #1 of SELECT list is not in GROUP BY clause and contains nonaggregated column 'manuscripts.folios.thumb_url' which is not functionally dependent on columns in GROUP BY clause; this is incompatible with sql_mode=only_full_group_by in /var/www/manuscriptsdev/home.php:172
Stack trace:
#0 /var/www/manuscriptsdev/home.php(172): PDOStatement->execute()
#1 /var/www/manuscriptsdev/index.php(183): include('...')
#2 {main}
  thrown in /var/www/manuscriptsdev/home.php on line 172

Fatal error: Uncaught PDOException: SQLSTATE[42000]: Syntax error or access violation: 1055 Expression #1 of SELECT list is not in GROUP BY clause and contains nonaggregated column 'manuscripts.folios.thumb_url' which is not functionally dependent on columns in GROUP BY clause; this is incompatible with sql_mode=only_full_group_by in /var/www/manuscriptsdev/home.php:172
Stack trace:
#0 /var/www/manuscriptsdev/home.php(172): PDOStatement->execute()
#1 /var/www/manuscriptsdev/index.php(183): include('...')
#2 {main}
```

**Comment by diedrick** [comment]
I think the answer to this one is to change a setting in MySQL, as listed in the 3rd answer here:
https://stackoverflow.com/questions/41887460/select-list-is-not-in-group-by-clause-and-contains-nonaggregated-column-inc

--

There is a system variable ONLY_FULL_GROUP_BY in MySql engine.

From Mysql Version 5.7.5: ONLY_FULL_GROUP_BY SQL mode is enabled by default

Before Version 5.7.5: ONLY_FULL_GROUP_BY was not enabled by default.

If the ONLY_FULL_GROUP_BY SQL mode is enabled (which it is by default from version 5.7.5), MySQL rejects queries for which the select list, HAVING condition, or ORDER BY list refer to non-aggregated columns that are neither named in the GROUP BY clause nor are functionally dependent on them.

To sort out the issue, use any one solution (out of below 3)
(1) PHPMyAdmin

Disable: ONLY_FULL_GROUP_BY mode

if you are using phpMyAdmin then change the sql_mode setting as mentioned in the below screenshot. enter image description here

Edit sql mode variable and remove the ONLY_FULL_GROUP_BY text from the value

OR

(2) SQL/Command prompt

Disable: ONLY_FULL_GROUP_BY mode by running the below command.

SET GLOBAL sql_mode=(SELECT REPLACE(@@sql_mode,'ONLY_FULL_GROUP_BY',''));
OR

(3) Don't use SELECT *

Do not disable the ONLY_FULL_GROUP_BY mode but

Use relevant column in SELECT query. relevant means columns, which are either coming in group by clause or column with the aggregate function (MAX, MIN, SUM, COUNT etc)

**Comment by dylan-k** [comment]
Of these, option 2 looks best. PHPMyAdmin isn't installed and option 3 looks too restrictive.

@diedrick Is option 2 something you're able to run or do you need an admin to do it?

**Comment by diedrick** [comment]
I had tried to run option 2, and I don't have permissions for that. I also don't yet have access to the MySQL via offsite like I have on the current side. 

Generally when setting up a server, I create a mysql offsiteadmin user that only has access via one or two IP addresses (easy to create users from specific locations) and give them most permissions. Don't give them those permissions locally, though, so no localhost/127.0.0.1, as that would be an easy privilege escalation. (I could set this up, too, if giving sudo access temporarily.)

BTW looks like there's a third option to disable ONLY_FULL_GROUP_BY via my.cnf, though I don't know if that changes anything for any other systems you have on the site.

My IP is 75.9.105.130.

**Comment by dylan-k** [comment]
ok I've let @jimgonzalezjr know about this and he should be able to get you the access you need.

**Comment by jimgonzalezjr** [comment]
Seems the query is not ANSI compliant, I fixed the statement by adding ANY_VALUE() to the select term for the thumb_url colum this makes the query compliant and runs perfectly.

**Comment by diedrick** [comment]
The query and the site were written many years ago without any ANSI requirements. We could solve this by a simple settings change, or we could rewrite every site query. I can't estimate the amount of time this will take, as the ANY_VALUE() trick didn't seem to solve the issue. Spending an hour on the query on line 107 of the ajax.php call, I can't get it to work without deeper tearing it apart, and I will have to reverse engineer what the function does in general from the working live site. Is this worth it? Why would the site not being ANSI compliant be an issue? 

Perhaps if you told me more of the reasoning behind why we shouldn't use the setting, I can find a more in-between solution.

As well, I still don't have access to the database to even try to test some of the assumptions.

**Comment by dylan-k** [comment]
@diedrick it might help to share some background, here. This new development environment is part of a larger project here at the Walters. We're upgrading each of our web servers and, when possible, merging them. Having the main collection website, the art site, and the API together opens up some opportunities for us, while also cutting down on the number of different servers we'd need to maintain.

With each of our sites, we've needed to do some adjustments, especially over on the art site. On that site, by the way, we do use `SELECT *`. So, if it works there, we shouldn't need to avoid it here. @jimgonzalezjr do you have any thoughts on why it works in one place and not another? I can show you those queries if it helps.

Some questions:
  - You mentioned, "the `ANY_VALUE()` trick didn't seem to solve the issue." Why not?
  - Are you able to connect to the database now? You should be able to configure a MySQL client to connect first by SSH and then to MySQL, or perhaps the open port is the fix you needed?

**Comment by diedrick** [comment]
@dylan-k Thanks for the more detailed context. I'm entirely for upgrading the site to a more standardized, more modern API feel. I just can't know how long that'll take, and I assume a budget for that would be limited. I'd personally think we'd do better to not try to fix just the ANSI issue specifically, but more by recreating the API calls -- that PHP and the MySQL statements are distinctly unmodern, not future friendly, and potentially could have security issues because of their age. 

Maybe because of the settings change, I'm seeing a different error now in the ajax call, the script is now printing out a $stmt, and that's what fails. However, I tried fix it, but the file permissions on that don't allow me to change that file anymore. If you don't want to rewrite the APIs now, the `ONLY_FULL_GROUP_BY` mode should be able to be set just for the `manuscripts` user, so you don't have to force it site wide.

As to why the `ANY_VALUE` system didn't work, it's a complex statement with many fields that need to be affected, it seems, which is why I needed to have :3306 access so I can investigate it in a parallel environment. The output error messaging didn't explain much, even when I tried adding an any_value to all the fields, so need to separate it out.

In terms of offiste MySQL access, trying the `manuscripts` user login, I get:
MySQL said: Authentication plugin 'caching_sha2_password' cannot be loaded: dlopen(/usr/local/mysql/lib/plugin/caching_sha2_password.so, 2): image not found

TL;DR: I can't save the /assets/ajax.php anymore via SSH/SFTP (permissions), and I still don't have offsite MySQL access.

**Comment by jimgonzalezjr** [comment]
I have fixed the issue with ajax.php. I was moving files around and I did not change ownership

 

As for the mysql error seem is a new authentication method and you need to update your client.  I changed to less secure  native.

 

alter user 'manuscripts'@'%' identified with mysql_native_password by ‘xxxxx’

 



 

If you have any questions please call my direct line 443-807-8076

 

Thanks 

Jim Gonzalez

IT Director

iMPACT Business Solutions

**Comment by dylan-k** [comment]
@jimgonzalezjr Michael has said that he can connect to the server, and its database, and that the file permissions seem to be in order. Thank you for your help with this.

@diedrick Now that you can connect and look around, let us know when you can, what's required to fix this issue and anything else needed as part of the server move.

Just as an update on the project, we're also going to test the admin scripts that publish to this site, to be sure they work in the new environment.
